File: modules/wem-api/src/main/java/com/enonic/wem/api/content/page/text/TextComponent.java
package com.enonic.wem.api.content.page.text;


import java.util.Objects;

import com.enonic.wem.api.content.page.AbstractPageComponent;
import com.enonic.wem.api.content.page.ComponentName;
import com.enonic.wem.api.content.page.PageComponentType;

public class TextComponent
    extends AbstractPageComponent
{
    private String text;

    protected TextComponent( final Builder builder )
    {
        super( builder );
        this.text = builder.text;
    }

    public static Builder newTextComponent()
    {
        return new Builder();
    }

    public PageComponentType getType()
    {
        return TextComponentType.INSTANCE;
    }

    public String getText()
    {
        return text;
    }

    @Override
    public boolean equals( final Object o )
    {
        if ( this == o )
        {
            return true;
        }
        if ( o == null || getClass() != o.getClass() )
        {
            return false;
        }
        if ( !super.equals( o ) )
        {
            return false;
        }

        final TextComponent that = (TextComponent) o;

        if ( text != null ? !text.equals( that.text ) : that.text != null )
        {
            return false;
        }

        return true;
    }

    @Override
    public int hashCode()
    {
        return Objects.hash( super.hashCode(), text );
    }

    public static class Builder
        extends AbstractPageComponent.Builder

    {
        private String text;

        private Builder()
        {
        }

        public Builder name( ComponentName value )
        {
            this.name = value;
            return this;
        }

        public Builder name( String value )
        {
            this.name = new ComponentName( value );
            return this;
        }

        public Builder text( String value )
        {
            this.text = value;
            return this;
        }

        public TextComponent build()
        {
            return new TextComponent( this );
        }
    }
}

File: modules/wem-admin-ui/src/main/resources/web/admin/common/js/content/page/text/TextComponent.ts
module api.content.page.text {

    import Region = api.content.page.region.Region;

    export class TextComponent extends api.content.page.Component implements api.Equitable, api.Cloneable {

        private text: string;

        constructor(builder?: TextComponentBuilder) {
            super(builder);
            if (builder) {
                this.text = builder.text;
            }
        }

        getText(): string {
            return this.text;
        }

        setText(value: string) {
            this.text = value;
        }

        toJson(): api.content.page.PageComponentTypeWrapperJson {

            var json: TextComponentJson = <TextComponentJson>super.toPageComponentJson();
            json.text = this.text != null ? this.text : null;

            return <api.content.page.PageComponentTypeWrapperJson> {
                TextComponent: json
            };
        }

        equals(o: api.Equitable): boolean {

            if (!api.ObjectHelper.iFrameSafeInstanceOf(o, TextComponent)) {
                return false;
            }

            var other = <TextComponent>o;

            if (!super.equals(o)) {
                return false;
            }

            if (!api.ObjectHelper.stringEquals(this.text, other.text)) {
                return false;
            }

            return true;
        }

        clone(generateNewPropertyIds: boolean = false): TextComponent {
            return new TextComponentBuilder(this).build();
        }
    }

    export class TextComponentBuilder extends api.content.page.PageComponentBuilder<TextComponent> {

        text: string;

        constructor(source?: TextComponent) {

            super(source);

            if (source) {
                this.text = source.getText();
            }
        }

        public fromJson(json: TextComponentJson, region: Region): TextComponentBuilder {

            if (json.text) {
                this.setText(json.text);
            }

            this.setName(new api.content.page.ComponentName(json.name));
            this.setParent(region);

            return this;
        }

        public setText(value: string): TextComponentBuilder {
            this.text = value;
            return this;
        }

        public build(): TextComponent {
            return new TextComponent(this);
        }
    }
}